nav2_collision_monitor/src/polygon.cpp

```diff
@@ -168,6 +168,34 @@ bool Polygon::isShapeSet()
   return true;
 }
 
+bool Polygon::isUsingPolygonVelocitySelector()
+{
+  return !polygon_velocity_.empty();
+}
+
+void Polygon::updatePolygonByVelocity(const Velocity & cmd_vel_in)
+{
+  for (auto & polygon : polygon_velocity_) {
+
+    if (polygon->isInRange(cmd_vel_in)) {
+      // Set the polygon that is within the speed range
+      poly_ = polygon->getPolygon();
+
+      // Update visualization polygon
+      polygon_.polygon.points.clear();
+      for (const Point & p : poly_) {
+        geometry_msgs::msg::Point32 p_s;
+        p_s.x = p.x;
+        p_s.y = p.y;
+        // p_s.z will remain 0.0
+        polygon_.polygon.points.push_back(p_s);
+      }
+
+      return;
```

Comment:
The first found PolygonVelocity will be applied. In case if we will have many overlapped polygons_velocity, responding to the same velocity, safest rule may not be found. However, this is the big question of Collision Monitor logic design, as the decision of making safest solution is made at `CollisionMonitorNode` level, which is above `Polygon` logic. If you will have any proper solution/suggestion for this situation, please check how could it be applied w/o large re-work of CM, or not? Otherwise, we have to assume that velocity ranges in {linear, angular} space should not intersect for one polygon.

Reply:
If we want to support finding the safest velocity for overlapping polygons without a major re-work, we can do something like this(but which polygon to visualise for `PolygonVelocity`?):

```cpp
    if (polygon->isUsingPolygonVelocitySelector()) {
      // Get all the polygons that is in range
      polygonsToProcess = polygon-> getPolygonsByVelocity(cmd_vel_in);
    } else {
      // Otherwise, dynamically update polygon coordinates when it necessary
      polygon->updatePolygon();
      polygonsToProcess.push_back(polygon);
    }

    for (auto & polygon: polygonsToProcess){
      const ActionType at = polygon->getActionType();
      if (at == STOP || at == SLOWDOWN || at == LIMIT) {
        // Process STOP/SLOWDOWN for the selected polygon
        if (processStopSlowdownLimit(polygon, collision_points, cmd_vel_in, robot_action)) {
          action_polygon = polygon;
        }
      } else if (at == APPROACH) {
        // Process APPROACH for the selected polygon
        if (processApproach(polygon, collision_points, cmd_vel_in, robot_action)) {
          action_polygon = polygon;
        }
      }
    }
```
I would agree that velocity ranges within one polygon shouldn't be overlapped, and if it does, we should choose the first polygon per the config.

Reply:
@kaichie, thanks for the update of your thoughts on it. The problem to draw the safest polygon could be resolved by adding a `for (auto sub_polygon : polygon->getPolygonsByVelocity(cmd_vel_in)) {...}` sub-cycle with constant `cmd_vel_in` as input to realize which sub-polygon is the best locally and then comparing produced by each polygon velocities globally to make the safest decision. We could switch to `std::vector<Velocity> generated_cmd_vel` per each polygon / sub-polygon and then select the best solution per each polygon though its sub-polygons, and globally. The knowledge about best polygon shape selected from sub-polygons could be stored (or re-used, it should be carefully checked - could we do this) in some field of `Polygon` class.

Another alternative I see - to move the best velocity selection `if() {processStopSlowdownLimit()} else if() {processApproach()}` logic into one separate routine, and then pass this routine callback to the polygon selector to compare each sub-polygon on its safest solution. The safest velocities should be stored again, in some `Velocity` vector. This is not straightforward solution, but also could take place.

Anyway, the main subject here is - which use-case might we cover by overlapping sets of different velocities for one polygon. I can not find any of them for now: the ticket initially was made to move out robot from some situations when CM was already stuck the robot, and we thus need to change the safety shape to move robot backward; or rotate. Or another case will use the robot's speed to change dynamically safety shape, but there is no intention to set multiple shapes per different velocities, that could work at the same time. For this, we have initial CM's multi-polygons mechanism, and thus, I do not see any reason why we need to double this logic into sub-polygons. In case if developers wants to have multiple choices for one robot, they could define multiple polygons (with its own velocity rules per each) , instead of intentionally making sub-polygons overlapped by velocities.

Anyway, if you have any example, when the above won't work, I think, we need to think about supporting safest selection through sub-polygons, as well.

Reply:
> However, the main subject here is - which use-case might we cover by overlapping sets of different velocities for one polygon. I can not find any of them for now: the ticket initially was made to move out robot from some situations when CM was already stuck the robot, and we thus need to change the safety shape to move robot backward; or rotate. 

@AlexeyMerzlyakov Yes this is the use-case for now, not aware of other scenario that would require finding the safest sub-polygons with overlapping velocity. However it is a good discussion to know that we can still do this, if needed, without a major re-work on the CM.